python/taichi/ad/_ad.py

```diff
@@ -85,14 +103,133 @@ def __exit__(self, _type, value, tb):
+                    np.float64)
+
+                self.reset(mode="restore")
+                x_pos(x, tangent_np, eps)
+                for func, args in self.calls:
+                    func(*args)
+                loss_pos = self.loss.to_numpy()
+
+                self.reset(mode="restore")
+                x_neg(x, tangent_np, eps)
+                for func, args in self.calls:
+                    func(*args)
+                loss_neg = self.loss.to_numpy()
+
+                ip_numerical = (loss_pos - loss_neg) * 0.5 / eps
+                x_grad_np = x.grad.to_numpy()
+                extra_dim = x_grad_np.ndim - tangent_np.ndim
+                if extra_dim == 1:
+                    tangent_np = np.expand_dims(tangent_np, axis=-1)
+                if extra_dim == 2:
+                    tangent_np = np.expand_dims(tangent_np, axis=-1)
+
+                ip_autodiff = np.sum(x_grad_np * tangent_np)
+                err = abs(ip_autodiff - ip_numerical)
+                # print("autodiff:", ip_autodiff)
+                # print("numerical:", ip_numerical)
+                if ip_numerical != 0:
+                    re = err / abs(ip_autodiff)
+                else:
+                    re = err / (abs(ip_autodiff) + 1e-20)
+                re_range.append(re)
+
+                if err * 100 < abs(ip_autodiff):
+                    check_pass = True
+                    break
+
+            self.result[i] = check_pass
+            plt.loglog(np.array(self.eps_range[:len(re_range)]),
+                       np.array(re_range),
+                       "-*",
+                       label="variable " + str(i))
+
+            if not check_pass:
+                print(i, "relative error:", min(re_range))
+            else:
+                print(i, "grad check pass")
+
+        if all(self.result):
+            print("all grad check pass")
+        else:
+            plt.xlabel("finite difference step size")
+            plt.ylabel("relative error")
+            plt.title("Grad check Error Convergence")
+            plt.legend()
+            plt.show()
+
+        self.reset(mode="restore")
+        for func, args in self.calls:
+            func(*args)
+
+    def reset(self, mode):
+        assert mode == "save" or mode == "restore"
+        impl.get_runtime().materialize()
+        if mode == "save":
+            self.backup_variables = []
+        else:
+            backup_variables = deepcopy(self.backup_variables)
+
+        def visit(node):
+            for _i in range(node.ptr.get_num_ch()):
+                ch = node.ptr.get_ch(_i)
+                if not ch.is_place():
+                    visit(SNode(ch))
+                else:
+                    if not ch.is_primal():
+                        continue
+                    if mode == "save":
+                        self.backup_variables.append(
+                            ScalarField(Expr(ch.get_expr())).to_numpy())
```

Comment:
Have we handled the lose precision warning yet ?

Reply:
Yes, it will automatically convert taichi type to corresponding numpy type